File: scripts/update_readme.py
from datetime import datetime, timedelta, timezone
import requests
import json
import re
import os
from collections import Counter

USERNAME = "RafaelaSommer"
README_PATH = "README.md"
SETTINGS_PATH = ".github/settings.json"
TOKEN = os.getenv("GITHUB_TOKEN")

BRAZIL_TZ = timezone(timedelta(hours=-3))
TZ_LABEL = "Horário de Brasília"

now_utc = datetime.now(timezone.utc)
now_brazil = now_utc.astimezone(BRAZIL_TZ)

def load_settings(path):
    if not os.path.exists(path):
        return {}
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
            return data if isinstance(data, dict) else {}
    except Exception:
        return {}

settings = load_settings(SETTINGS_PATH)

headers = {"Accept": "application/vnd.github+json"}
if TOKEN:
    headers["Authorization"] = f"Bearer {TOKEN}"

# --- Buscar repositórios ---
repos = []
page = 1
while True:
    response = requests.get(
        f"https://api.github.com/users/{USERNAME}/repos?per_page=100&page={page}",
        headers=headers
    )
    response.raise_for_status()
    batch = response.json()
    if not batch:
        break
    repos.extend(batch)
    page += 1

# --- Contagem total de projetos ---
total_projects = len(repos)

# --- Contar linguagens ---
languages = Counter()
for repo in repos:
    if repo.get("language"):
        languages[repo["language"]] += 1

lang_lines = "<br>\n".join(
    f"• {lang}: {count}" for lang, count in languages.most_common()
) if languages else "• Nenhuma linguagem detectada"

last_update = now_brazil.strftime("%d/%m/%Y %H:%M")

# --- Próxima atualização fixa (12h ou 19h) ---
today = now_brazil.date()

update_12 = datetime.combine(today, datetime.min.time()).replace(
    hour=12, tzinfo=BRAZIL_TZ
)
update_19 = datetime.combine(today, datetime.min.time()).replace(
    hour=19, tzinfo=BRAZIL_TZ
)

if now_brazil < update_12:
    next_update = update_12
elif now_brazil < update_19:
    next_update = update_19
else:
    next_update = update_12 + timedelta(days=1)

next_update_str = next_update.strftime("%d/%m/%Y %H:%M")

# --- Atualizar settings.json ---
settings.update({
    "username": USERNAME,
    "total_projects": total_projects,
    "languages": dict(languages),
    "last_update": last_update,
    "next_update": next_update_str,
    "updated_by": "github-actions"
})

os.makedirs(".github", exist_ok=True)
with open(SETTINGS_PATH, "w", encoding="utf-8") as f:
    json.dump(settings, f, indent=2, ensure_ascii=False)

# --- Criar bloco para README ---
info_block = (
    "<!-- INFO-START -->\n"
    "📌 <strong>Últimas Atualizações</strong><br>\n"
    f"📊 <strong>Total de projetos:</strong> {total_projects}<br>\n"
    "🧠 <strong>Projetos por linguagem:</strong><br>\n"
    f"{lang_lines}<br>\n"
    "🔄 <strong>Atualização automática:</strong> a cada 10 minutos via GitHub Actions<br>\n"
    "🔮 <strong>Próxima atualização prevista:</strong><br>\n"
    f"• {next_update_str} ({TZ_LABEL})<br>\n"
    f"⏱️ <strong>Última atualização:</strong> {last_update} ({TZ_LABEL})\n"
    "<!-- INFO-END -->"
)

# --- Atualizar README ---
with open(README_PATH, "r", encoding="utf-8") as f:
    content = f.read()

pattern = r"<!-- INFO-START -->.*?<!-- INFO-END -->"

if re.search(pattern, content, flags=re.DOTALL):
    content = re.sub(pattern, info_block, content, flags=re.DOTALL)
else:
    content += "\n\n" + info_block

with open(README_PATH, "w", encoding="utf-8") as f:
    f.write(content)

File: scripts/update_readme.js
#!/usr/bin/env node

require("dotenv").config()

const fs = require("fs")
const path = require("path")
const axios = require("axios")
const { DateTime } = require("luxon")

const ROOT = path.join(__dirname, "..")

const SETTINGS = JSON.parse(
  fs.readFileSync(path.join(ROOT, ".github/settings.json"), "utf8")
)

const USER = SETTINGS.github_user
const TIMEZONE = SETTINGS.timezone
const INTERVAL = SETTINGS.interval_minutes

const TOKEN = process.env.GITHUB_TOKEN

if (!TOKEN) {
  console.error("❌ GITHUB_TOKEN não encontrado")
  process.exit(1)
}

async function fetchGitHub() {
  try {
    const res = await axios.post(
      "https://api.github.com/graphql",
      {
        query: `
        query {
          user(login:"${USER}") {
            followers { totalCount }
            repositories(first:100) {
              nodes {
                stargazerCount
              }
            }
          }
        }`
      },
      {
        headers: {
          Authorization: `Bearer ${TOKEN}`
        }
      }
    )

    return res.data.data.user

  } catch (err) {
    console.error("❌ Erro ao buscar dados do GitHub")

    if (err.response) {
      console.error("Status:", err.response.status)
      console.error("Resposta:", err.response.data)
    } else {
      console.error(err.message)
    }

    process.exit(1)
  }
}

function updateReadme(dynamicContent) {

  const readmePath = path.join(ROOT, "README.md")

  let content = fs.readFileSync(readmePath, "utf8")

  const start = "<!--START_SECTION:dynamic-->"
  const end = "<!--END_SECTION:dynamic-->"

  const regex = /<!--START_SECTION:dynamic-->[\s\S]*<!--END_SECTION:dynamic-->/

  const newBlock = `${start}
${dynamicContent.trim()}
${end}`

  if (!regex.test(content)) {
    console.error("❌ Bloco dinâmico NÃO encontrado no README")
    console.error("👉 Verifique se existe exatamente:")
    console.error(start)
    console.error(end)
    process.exit(1)
  }

  // 🔥 Remove seção "IMPORTANTE" de forma robusta (somente no README final)
 content = content.replace(
  /<!--START_IMPORTANTE-->[\s\S]*?<!--END_IMPORTANTE-->/g,
  ''
).trim()

  const updated = content.replace(regex, newBlock)

  fs.writeFileSync(readmePath, updated, "utf8")

  console.log("✅ README atualizado com sucesso (sem seção IMPORTANTE)")
}

async function main() {

  const now = DateTime.now().setZone(TIMEZONE)
  const next = now.plus({ minutes: INTERVAL })

  const user = await fetchGitHub()

  const repos = user.repositories.nodes

  const followers = user.followers.totalCount
  const stars = repos.reduce((a, r) => a + r.stargazerCount, 0)

  const dynamicContent = `
## 🔄 Atualização Automática

🕒 Última atualização:  
${now.toFormat("dd/MM/yyyy HH:mm:ss")} (Horário de Brasília)

🔁 Próxima atualização automática:  
${next.toFormat("dd/MM/yyyy HH:mm:ss")} (Horário de Brasília)

📊 **Followers:** ${followers}  
⭐ **Stars:** ${stars}
`

  updateReadme(dynamicContent)
}

main()